docs/changelog/4_x.rst

```diff
@@ -0,0 +1,467 @@
+====
+v4.0
+====
+
+:edb-alt-title: EdgeDB v4
+
+To play with the new features, install the CLI using `our installation guide
+<https://www.edgedb.com/install>`_ and initialize a new project.
+
+.. code-block:: bash
+
+  $ edgedb project init --server-version 4.0-beta.1
+
+
+Upgrading
+=========
+
+**Local instances**
+
+To upgrade a local project, first ensure that your CLI is up to date with
+``edgedb cli upgrade``. Then run the following command inside the project
+directory.
+
+.. code-block:: bash
+
+  $ edgedb project upgrade --to-testing
+
+Alternatively, specify an instance name if you aren't using a project.
+
+.. code-block:: bash
+
+  $ edgedb instance upgrade --to-testing -I my_instance
+
+**Hosted instances**
+
+To upgrade a remote (hosted) instance, we recommend the following
+dump-and-restore process.
+
+1. Spin up an empty 4.0 instance. You can use one of our :ref:`deployment
+   guides <ref_guide_deployment>`, but you will need to modify some of the
+   commands to use our testing channel and the beta release.
+
+   Under Debian/Ubuntu, when adding the EdgeDB package repository, use this
+   command instead:
+
+   .. code-block:: bash
+
+       $ echo deb [signed-by=/usr/local/share/keyrings/edgedb-keyring.gpg] \
+           https://packages.edgedb.com/apt \
+           $(grep "VERSION_CODENAME=" /etc/os-release | cut -d= -f2) main \
+           | sudo tee /etc/apt/sources.list.d/edgedb.list
+
+   Use this command for installation under Debian/Ubuntu:
+
+   .. code-block:: bash
+
+       $ sudo apt-get update && sudo apt-get install edgedb-4-beta1
+
+   Under CentOS/RHEL, use this installation command:
+
+   .. code-block:: bash
+
+       $ sudo yum install edgedb-3
+
+   In any required ``systemctl`` commands, replace ``edgedb-server-3`` with
+   ``edgedb-server-4``.
+
+   Under any Docker setups, supply the ``4.0-beta.1`` tag.
+
+2. Take your application offline, then dump your v3.x database with the CLI
+
+   .. code-block:: bash
+
+       $ edgedb dump --dsn <old dsn> --all --format dir my_database.dump/
+
+   This will dump the schema and contents of your current database to a
+   directory on your local disk called ``my_database.dump``. The directory name
+   isn't important.
+
+3. Restore the empty v4.x instance from the dump
+
+   .. code-block:: bash
+
+       $ edgedb restore --all my_database.dump/ --dsn <new dsn>
+
+   Once the restore is complete, update your application to connect to the new
+   instance.
+
+   This process will involve some downtime, specifically during steps 2 and 3.
+
+**Pre-1.0 Instances**
+
+If you're still running pre-1.0 EdgeDB instances (e.g., 1.0-beta3) and want to
+upgrade to 4.0, we recommend you upgrade to version 2.x first, followed by
+another upgrade to 4.0, both using the same dump-and-restore process.
```

Comment:
this is copied from the 3.0 changelog with `s/3.0/4.0/` - is going directly from 2.0 -> 4.0 supported? or is an intermediate upgrade to 3.0 required?

Reply:
I don't think I've tested the direct 2.0 -> 4.0 upgrade path.

Reply:
We test dump/restores from 1.4, 2.0, a pre-3.0 dev dump, and 3.0, currently.

The thing that prompted this was that we removed a bunch of hacks for supporting various pre-1.0 dumps.